### ISSUE TITLE
[@types/express-serve-static-core] async handlers typings

### ISSUE THREAD

**Issue by jlowcs** [DefinitelyTyped/DefinitelyTyped#50871]
I tried bumping `@types/express-serve-static-core` to the latest version but the [changes to the return type of handlers](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/7a41e02d90d3c0cc02104272a28ac4d89b5d64d4#diff-5962769e8693df44c9e1083881fb1da149a67262026db939b7430536d514e7af) (to `void`) introduced a typing error for async handlers.

Should we instead go with a `void | Promise<void>` return type?

**Comment by AviVahl** [comment]
Does express actually accept a promise (aka handles rejections)? If not, which I suspect is the case, `Promise<void>` should not be accepted, as that could potentially lead to unhandled promise rejections.

**Comment by jlowcs** [comment]
In express 4, async functions are supported, but thrown errors are ignored (you gotta catch them and call `next(err)`)

In express 5, they are completely supported (see

**Comment by AviVahl** [comment]
> In express 4, async functions are supported, but thrown errors are ignored (you gotta catch them and call `next(err)`)

If nothing handles rejections of Promises in `express@4`, like you're describing, then request handlers shouldn't be explicitly allowed to return a Promise. This makes sense to me. This is also only way for tooling such as eslint (with typescript-eslint) could catch cases when Promise is returned in non Promise-handling context. 

> In express 5, they are completely supported (see https://expressjs.com/en/guide/writing-middleware.html)

Finally. 5 is an an alpha release and I see `@types/express` didn't update to it yet. That's _probably_ what should happen. Creating a "v4" folder containing the current types and updating the API to the changes in `express@5`. I'm not sure what's the official position of DT regarding this, as the v5-alpha release not considered stable and is probably not widely used.

**Comment by sixmen** [comment]
yes, express 4 does not handle Promise. But I want to use async/await syntactic sugar, and it returns `Promise<void>`.

```
router.get('/hello', async (req, res) => {
  try {
    await doSomething();
    res.send('Hello');
  } catch (error) {
    res.send(error.message);
  }
});
```

Is there another solution?

**Comment by AviVahl** [comment]
@sixmen If you're manually handling possible promise rejections, then a possible workaround you can cast the type of the request handler to what `router.get` expects. (`express.Handler/RequestHandler`?)

**Comment by maggie0002** [comment]
Express 5 is now in beta, and looking pretty stable. I had run in to the typings error (https://github.com/expressjs/express/issues/4892). What is the best way to go about updating, is it pending a community contribution of a Express 5 typings?

**Comment by jamiehaywood** [comment]
@maggie0002 I've written my own type wrapper for middlewares, see my SO answer here: https://stackoverflow.com/questions/34508081/how-to-add-typescript-definitions-to-express-req-res/67560007#67560007

For completeness here's the code below:

```typescript
export type ExpressMiddleware<ReqBody = AnyObject, Res = AnyObject | string, QueryString = AnyObject> = (
  req: TypedRequest<ReqBody, QueryString>,
  res: Response<Res>,
  next: NextFunction,
) =>
  | Promise<void | Response<Res>>
  | Promise<void>
  | ExpressMiddleware<ReqBody, Res, QueryString>
  | Response<any, Record<string, any>>
  | void
```

then consume like so:

```typescript
import { ExpressMiddleware } from '../interfaces/ExpressMiddleware';

type Req = { email: string; password: string };

type Res = { message: string };

export const signupUser: ExpressMiddleware<Req, Res> = async (req, res) => {
  /* strongly typed `req.body`. yay autocomplete 🎉 */
  res.json({ message: 'you have signed up' }) // strongly typed response obj
};
```

**Comment by malik-samad** [comment]
I resolved this issue by using express RequestHandler type for the handlers. like this
```typescript
accountRoutes.post('/endpoint', (async (req, res) => {
    // handler code here 
}) as RequestHandler);
```

**Comment by NatoBoram** [comment]
But then you're losing type safety. This functions returns a `Promise<void>`, not just `void`. Not that it matters much, but it's ignoring the real issue. There should be a way to override `RequestHandler`.

**Comment by FelixZY** [comment]
Looks like the fix was reverted in #69845 and moved to #69846 . Reopen?
